050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,801 @@
+>        3. If deletion failed with any other error this is reported via a condition in the `status.conditions` and the finalizer is not removed.
+---
+
+Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
+This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.
+This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.
+
+> An alternative to `spec.managed` would be "explicit deletion" via a `spec.delete` flag.
+> This alternative was rejected because it's incompatible with gitops.
+
+---
+
+#### Example 4: `KafkaTopic` deletion WITHOUT propagation to Kafka
+
+> 1. Assume the topic is currently Kube-managed
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 123
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 123 # == generation => it is Kube-managed
+>     ```
+> 
+> 2. The user changes the new `spec.managed` field to `False` so that the topic is no longer Kube-managed.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       managed: False # do not synchronize to Kafka
+>     status:
+>       observedGeneration: 123 # != generation => it is Kube-managed
+>     ```
+> 
+> 3. The user waits until the operator has observed that change, indicated by the `status.observedGeneration` matching the `metadata.generation`:
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       managed: False # do not synchronize to Kafka
+>     status:
+>       observedGeneration: 124 # == generation => it is now Kube-managed
+>     ```
+> 
+> 4. The user can then delete the resource.
+>     Because of the presence of a `metadata.finalizer` the resource is not actually deleted.
+>     Instead Kube adds the `metadata.deletionTimestamp` field.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>       deletionTimestamp: 20230301T000000.000
+>     spec:
+>       managed: False # do not synchronize to Kafka
+>     status:
+>       observedGeneration: 124 # == generation => it is not syncrhonized
+>     ```
+> 
+> 5. The operator notices the `metadata.deletionTimestamp` field.
+>     Since the `spec.magnaged` is `False` it does **not** attempt topic deletion.
+>     1. It removes `strimzi.io/topic-operator` from the `metadata.finalizer`.
+>     2. Kube proceeds to remove the resource.
+
+Note that `spec.managed=False` means that a `KafkaTopic` can exist without there being a corresponding topic in Kafka.
+
+---
+
+
+### Multi-namespace support
```

Comment:
The proposal mostly looks good to me. But I think that this is a bit premature to be included in it. I think the multi-namespace support needs to be considered and designed in the context of the UO as well.

* It would be weird to have multi-namespace TO and not UO => it should be probably added in parallel
* It would be good to consider how the namespacing is done in TO and UO. In the UO, in the past, we considered following Kubernetes and their naming principles by suffixing the user name with the namespace as a suffix. This approach might not be suitable for TO and we might intentionally choose a different approach for each of them. But we should do so intentionally after consideration and not arrive at it randomly.
* The policy seems to introduce a parallel security aspect (parallel to Kafka ACLs / UO and to Kube RBAC). It seems to be very limited (managing topics through KafkaTopic resources) while not impacting in any way the other security parts - especially the Kafka ACLs. So, a `team.a` is seemingly an owner of topics `foo-app.*`. But `team.b` can easily create users to produce or consumer to / from these topics. It can even create users with ACLs to delete or modify them. I think this would need to be properly thought through and linked to the UO and ACL management to create some value. Doing it without it can prevent doing this in the future without compatibility issues.

That all seems to be way above this proposal. I would probably prefer to do one of the following:
* Remove the multi-namespace support from this proposal and leave it for later
* Adapt the multi-namespace support to give us more options for the future. E.g.
    * Remove the policy concept until it is properly thought through and deal with multi-namespace conflicts in the same way as for single-namespace conflicts for the time being.
    * Acknowledge the possible future differences how UO might handle multi-namespace support.

----

Assuming you would want to proceed with this as is, I think it is missing clarification on how the multi-namespace watching will be configured. And what the _default_ policy will be. Does the policy configure that? Does the example policy below mean that only namespaces `team.a`, `team.b` and `kafka.cluster.admins` will be watched? That would be inefficient as the number of watches would not scale well when using per-namespace watch.

Reply:
Perhaps calling it a tenancy model would be more matching than a security model as I referred to it in the initial comment. But I think the core of the comment stands.